### ISSUE TITLE
when starting multiple tasks using `consul lock` should be an option

### ISSUE THREAD

**Issue by kneufeld** [hashicorp/nomad#6367]
To prevent a thundering herd it would be great if nomad supported `consul lock` in some fashion so that I could upgrade a hundred jobs in a controlled fashion. A key feature would be to specify the key so that you could lock _across_ different jobs (as well as within a job/group if desired).

I'm aware of `splay` but that's a sub-optimal/nondeterministic solution. Is `1h` long enough? What if my cluster can handle three simultaneous job upgrades and doing all of them would take 20 minutes?

If the `pre_hook`, `post_hook` feature was implemented then maybe a user could hack something together using them. Or this feature could make use of them? My 2 cents on implementation. 😉

**Comment by tgross** [comment]
Hi @kneufeld and thanks for opening this issue!

I was wondering if you could provide a little more detail on what kind of controls you're looking for here. Are there scenarios that can't be met by using an [`upgrade`](https://www.nomadproject.io/docs/job-specification/update.html) block that you could describe?

**Comment by kneufeld** [comment]
Let's say I have 100 hundred customers, each would need their own nomad job file. If I push a new docker image then I'd have 100 concurrent upgrades, or I would need to write some sort of script that did one at a time and checked for health and version etc.

`upgrade` wouldn't work because each client would be in their own job not all within a single group. Actually... if `upgrade` had a new `get_consul_lock` that would totally work if all jobs used the same one.

I feel like I'm not explaining myself very well, please ask more questions.

**Comment by tgross** [comment]
Ok, I think I understand: Nomad has plenty of knobs for controlling upgrades/rollouts for a single Nomad job, but not between multiple unrelated Nomad jobs. This is by design -- the job is the unit of scheduling. But I can see the reason you might be concerned about this if your job deployments could cause a lot of extra resource usage (ex. bandwidth pulling down Docker images).

A possible work around is to use a `template` block that includes a [`key`](https://github.com/hashicorp/consul-template#key). This will place the allocation without starting the container, and then block until the key appears in Consul. This job will be marked "pending". You'd then just need a way to add the key to Consul to "unlock" the deploy. 

Here's our example job with a template block like that:

```hcl
job "example" {
  datacenters = ["dc1"]

  group "cache" {
    task "redis" {
      driver = "docker"

      config {
        image = "redis:3.2"

        port_map {
          db = 6379
        }
      }

      template {
        data = <<EOH
{{ key "lock" }}
EOH
        destination   = ".lock"
        change_mode   = "restart"
      }

      resources {
        cpu    = 500
        memory = 256

        network {
          mbits = 10
          port  "db"  {}
        }
      }
    }
  }
}
```

Note this will cause the space to be reserved for the job, so that might not be what you want. In the past when I've run a production Nomad cluster, I've written a little service/script to serialize deployments between jobs and check on their health before continuing. It's a similar situation to what you're referring but in our case it had more to do with human bandwidth to deal with our sometimes flaky applications 😀 

Hope this helps!

**Comment by kneufeld** [comment]
Thanks for the detailed response but if I have two near identical job files

```
job "client1" {
...
}}}
```

```
job "client2" {
...
}}}
```

Then changing `lock` will cause both jobs to restart at the same time. If I wrote my own script to wrap calls to `nomad` and it does it's own health checking that would  work but... `nomad` is an orchestrator, it would be nice if orchestrated.

I'm not a nomad dev so don't know all the internals and all the repercussions but adding a `get_lock` to `update` sure seems like a simple feature.

Anyhow, this was my request but if it's not feasible that's fine, I'll still love and use the hashistack. Feel free to close.

**Comment by tgross** [comment]
That's true, you'd need some kind of external coordination to do that. Which, as you say, would be nice if we could do in Nomad.

So for now I can give you two take-aways:
* There's upcoming feature work being done on "pre start" tasks that might fit the bill for you. I can't give you a timeline but the groundwork is being actively worked on.
* The ability to have cross-job coordination in a more general sense sounds like a good enhancement but will need some design work. I'm going to keep this issue as an open enhancement request for further discussion.

Thanks for explaining to me in such detail what you're trying to do here @kneufeld. It's a huge help!

**Comment by markmartirosian** [comment]
This would be great for running "consul-terraform-sync" and HPC workloads.
